gax-grpc/src/main/java/com/google/api/gax/grpc/OperationFuture.java

```diff
@@ -29,58 +29,117 @@
  */
 package com.google.api.gax.grpc;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.core.ApiFunction;
 import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
 import com.google.api.core.BetaApi;
+import com.google.api.gax.retrying.RetryingFuture;
 import com.google.longrunning.Operation;
 import com.google.protobuf.Message;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Executor;
+import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
 
 /** An ApiFuture which polls a service through OperationsApi for the completion of an operation. */
 @BetaApi
-public final class OperationFuture<ResponseT extends Message> implements ApiFuture<ResponseT> {
+public final class OperationFuture<ResponseT extends Message, MetadataT extends Message>
+    implements ApiFuture<ResponseT> {
+  @Override
+  public String toString() {
+    return pollingFuture.toString();
+  }
+
+  private final Object lock = new Object();
+
+  private final RetryingFuture<Operation> pollingFuture;
   private final ApiFuture<Operation> initialFuture;
   private final ApiFuture<ResponseT> resultFuture;
+  private final ApiFunction<Operation, MetadataT> metadataTransformer;
+
+  private volatile ApiFuture<Operation> peekedAttemptResult;
+  private volatile ApiFuture<MetadataT> peekedPollResult;
+  private volatile ApiFuture<Operation> gottenAttemptResult;
```

Comment:
Probably, but the naming is inspired by the naming of the method the object is returned from: peeked returned from peek() method, gotten returned from get() method.